### PR TITLE
[SYCL][CUDA] Test cases for bfloat16 math/elem wise joint_matrix

### PR DESCRIPTION
**Pull request by JackAKirk**
requires https://github.com/intel/llvm/pull/5964

bfloat16_builtins.cpp covers the bfloat16 scalar math function cases introduced by https://github.com/intel/llvm/pull/5964, using the tests from https://github.com/intel/llvm-test-suite/pull/897 (that cover all "storage type" uint16_t impl cases).

elem_wise_all_ops_cuda.cpp covers the portable elem wise ops using `wi_data`. Since CUDA does not support `joint_matrix_store` for certain data types that are only used in a/b type matrices, such as bfloat16 and int8, it is necessary to perform a `joint_matrix_mad` operation and then call `joint_matrix_store` on the accumulator matrix in order the reach the host code check.
Intel backend devices could still use this test in the future provided that a backend check is introduced. Ideally both backends could eventually use the same test code.

Signed-off-by: jack.kirk <jack.kirk@codeplay.com>

### CODE REVIEW

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (steffenlarsen) [inline SYCL/Matrix/joint_matrix_tensorcore.cpp]: Instead of relying on CI, could the test check the SM version of the device during execution and skip if it isn't SM80?

Additionally, does all of this test require SM80 or is it just a subset? If it is just a subset, it would be better to split it up so CI can at least test the parts that are expected to work with the current hardware.

Comment (JackAKirk) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: > Instead of relying on CI, could the test check the SM version of the device during execution and skip if it isn't SM80?

Yeah thanks! Conveniently `info::device::backend_version` has recently been implemented for CUDA
 
> Additionally, does all of this test require SM80 or is it just a subset? If it is just a subset, it would be better to split it up so CI can at least test the parts that are expected to work with the current hardware.

Yep thanks again. I've done this now.

**Comment by JackAKirk** [comment]
I have decided to mark this and the associated https://github.com/intel/llvm/pull/5964 as draft: I've have been working on the follow up to this: implementing vector of bfloat16 cases: the vector of bfloat16 cases could motivate a change in the joint_matrix bfloat16 implementation. As such I want to explore this further before merging the bfloat16 scalar implementation.

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (steffenlarsen) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: Thank you! Remind me, when does the CUDA backend fail if the device does not have the capabilities required for a kernel? Is it during build or is it when the specific kernel is enqueued?

If it's the former, the test may still fail despite the checks as the kernels may be put in the same device code module. There are a handful of ways to work around it:

1. Split the test into separate test files. Optionally they could share a common header.
2. Use macros in this test and compile and run each variant.
3. Add `-fsycl-device-code-split=per_kernel`. This will result in separate device code for each kernel, which may be a little aggressive here.

**Comment by steffenlarsen** [comment]
> I have decided to mark this and the associated [intel/llvm#5964](https://github.com/intel/llvm/pull/5964) as draft: I've have been working on the follow up to this: implementing vector of bfloat16 cases: the vector of bfloat16 cases could motivate a change in the joint_matrix bfloat16 implementation. As such I want to explore this further before merging the bfloat16 scalar implementation.

Sounds good! Keep me posted. 😄

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (JackAKirk) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: > Thank you! Remind me, when does the CUDA backend fail if the device does not have the capabilities required for a kernel? Is it during build or is it when the specific kernel is enqueued?
> 

It's the latter. I've ran the lit test on a lower SM arch than the tests support and it will currently skip the tests and pass.

Comment (steffenlarsen) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: Thanks for clarifying!

File: SYCL/BFloat16/bfloat16_builtins.cpp
Comment (dkhaldi) [inline SYCL/BFloat16/bfloat16_builtins.cpp]: should this be changed to sycl/sycl.hpp like the changes you made in #5964?

File: SYCL/BFloat16/bfloat16_builtins.cpp
Comment (dkhaldi) [inline SYCL/BFloat16/bfloat16_builtins.cpp:21]: Why do we still need this function in the bfloat16 case? this should use to_float function.

File: SYCL/Matrix/element_wise_all_ops_cuda.cpp
Comment (dkhaldi) [inline SYCL/Matrix/element_wise_all_ops_cuda.cpp:75]: As we discussed, it will be good to have the two APIs (get_wi_data and wi_marray) used in the same test and show that they deliver the same performance and IR/assembly. This test can be added here or in intel/llvm tests

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (dkhaldi) [inline SYCL/Matrix/joint_matrix_tensorcore.cpp]: Is this supported for int8_t and half as well?

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (dkhaldi) [inline SYCL/Matrix/joint_matrix_tensorcore.cpp]: I would prefer to keep this test for joint_matrix initial primitives: load, store, mad.
Then, you added a test for element wise operations in SYCL/Matrix/element_wise_all_ops_cuda.cpp
For the wi_marray feature, I would suggest to create a separate test for that and the math functions that users can do besides joint_matrix. Also, specify the types that work with these math functions.

**Review by dkhaldi — CHANGES_REQUESTED** [review]

File: SYCL/BFloat16/bfloat16_builtins.cpp
Comment (JackAKirk) [reply SYCL/BFloat16/bfloat16_builtins.cpp:21]: Thanks. I've switched to the implicit cast to float.

File: SYCL/BFloat16/bfloat16_builtins.cpp
Comment (JackAKirk) [reply SYCL/BFloat16/bfloat16_builtins.cpp]: Thanks, done.

File: SYCL/Matrix/element_wise_all_ops_cuda.cpp
Comment (JackAKirk) [reply SYCL/Matrix/element_wise_all_ops_cuda.cpp:75]: I've now added such a test named element_wise_wi_marray.cpp. I checked that the same assembly is output if we use/operate on either `sub_a.get_wi_data()[i]` or `sub_a.wi_data[i]`.

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (JackAKirk) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: full sycl 2020 marray support including all genfloat types: half, float, double has been added in the open PR here: https://github.com/intel/llvm/pull/6038. integer types are not supported by this function.

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (JackAKirk) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: Within the scope of this PR only `bfloat16` type is supported for the math functions tested in element_wise_all_ops_cuda.cpp. Test coverage for other types with support added in https://github.com/intel/llvm/pull/6038 is provided in https://github.com/intel/llvm-test-suite/pull/1002.

File: SYCL/BFloat16/bfloat16_builtins.cpp
Comment (dkhaldi) [reply SYCL/BFloat16/bfloat16_builtins.cpp:21]: @asudarsa, have we decided to use implicit cast rather than to_float?

Comment (JackAKirk) [reply SYCL/BFloat16/bfloat16_builtins.cpp:21]: `bfloat16.to_float(uint16_t raw)` as it is currently defined takes a raw storage type, and is a member function of bfloat16 rather than a free function: if there were to be an explicit cast I think there should be a free function added that can take bfloat16 rather than uint16_t.

File: SYCL/Matrix/joint_matrix_tensorcore.cpp
Comment (dkhaldi) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: I think the portion:
if constexpr (std::is_same<T1, bfloat16>::value) {
                marray<bfloat16, sub_a.wi_marray.size()> b, c;
                b = 2;
                c = 1;
                sub_a.wi_marray =
                    sycl::ext::oneapi::experimental::fma(sub_a.wi_marray, b, c);
              }
should move to a new test like joint_matrix_tensorcore_mathtest.
there, you can specify that this only works for bf16

Comment (JackAKirk) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: In the last commit I completely removed this code (as you suggested just leaving the initial joint_matrix primitives) and replaced it with element_wise_wi_marray.cpp which does something similar except with `fabs`: I could add all operations to element_wise_wi_marray.cpp if you think that is better, but I thought the purpose of it was to demonstrate the corresponding element wise math operations on the cuda specific `joint_matrix.wi_marray` and the general `joint_matrix.get_wi_data()[]` routes.

Comment (dkhaldi) [reply SYCL/Matrix/joint_matrix_tensorcore.cpp]: Sorry, I missed that you removed it. Both tests look good now.

**Review by steffenlarsen — DISMISSED** [review]
LGTM. @dkhaldi - Have your concerns been addressed?

**Comment by JackAKirk** [comment]
I can use the `ext_oneapi_bfloat16` aspect in this PR now that https://github.com/intel/llvm/pull/5720 has been merged so that there is not a cuda specific compute capability check blocking future generalization of the tests to other backends.  I will update this shortly.

**Comment by JackAKirk** [comment]
> I can use the `ext_oneapi_bfloat16` aspect in this PR now that [intel/llvm#5720](https://github.com/intel/llvm/pull/5720) has been merged so that there is not a cuda specific compute capability check blocking future generalization of the tests to other backends. I will update this shortly.

I've done this now. Note that I only made this change to the bfloat16_builtins.cpp test file. I've left element_wise_wi_marray.cpp as it is because it is only expected to be supported by cuda in the future.

**Review by dkhaldi — DISMISSED** [review]
LGTM

File: SYCL/BFloat16/bfloat16_builtins.cpp
Comment (steffenlarsen) [inline SYCL/BFloat16/bfloat16_builtins.cpp:1]: Now that it uses aspects, is this still needed?

Comment (JackAKirk) [reply SYCL/BFloat16/bfloat16_builtins.cpp:1]: I'm not actually sure. Maybe it isn't needed. Probably I can only find out by removing it and running the CI to test all backends. I'll push a commit removing it and see what happens.

Comment (steffenlarsen) [reply SYCL/BFloat16/bfloat16_builtins.cpp:1]: It appears that it fails to compile for other backends. Let us reintroduce this requirement. Could you please make a note specifying that it currently fails for other backends?

**Comment by steffenlarsen** [comment]
@JackAKirk - Looks like CI choked for some reason. Failures in verification seem unrelated. Would you mind pushing a merge-commit to retrigger CI, just to make sure it was a one-time hiccup.

**Comment by JackAKirk** [comment]
> @JackAKirk - Looks like CI choked for some reason. Failures in verification seem unrelated. Would you mind pushing a merge-commit to retrigger CI, just to make sure it was a one-time hiccup.

Sure: it seems that you can retrigger CI by closing and reopening the PR: seems a bit easier than merging the main branch.

**Comment by steffenlarsen** [comment]
Failures are unrelated. Merging this.

**Comment by steffenlarsen** [comment]
@JackAKirk - The tests seem to be failing in intel/llvm CI (see for [7131271000](https://github.com/intel/llvm/runs/7131271000?check_suite_focus=true) and [7131035571](https://github.com/intel/llvm/runs/7131035571?check_suite_focus=true)). Could you please address these?

**Comment by JackAKirk** [comment]
> @JackAKirk - The tests seem to be failing in intel/llvm CI (see for [7131271000](https://github.com/intel/llvm/runs/7131271000?check_suite_focus=true) and [7131035571](https://github.com/intel/llvm/runs/7131035571?check_suite_focus=true)). Could you please address these?

Looking into it now.

**Comment by JackAKirk** [comment]
> @JackAKirk - The tests seem to be failing in intel/llvm CI (see for [7131271000](https://github.com/intel/llvm/runs/7131271000?check_suite_focus=true) and [7131035571](https://github.com/intel/llvm/runs/7131035571?check_suite_focus=true)). Could you please address these?

I still don't understand these failure or why they are only occurring on the CI for new PRs.  I've explicitly added headers and used extension namespaces here https://github.com/intel/llvm-test-suite/pull/1072. Since these tests are passing for me locally I can't say for sure that this will fix it. I could also mark them XFAIL temporarily. It would be useful to see if https://github.com/intel/llvm-test-suite/pull/1072 fixes it though.

**Comment by pvchupin** [comment]
@JackAKirk, @steffenlarsen, I couldn't reproduce on other cuda machine, but when ran on the runner got a bunch of issues:

```
Failed Tests (11):
  SYCL :: Assert/assert_in_kernels.cpp
  SYCL :: Assert/assert_in_multiple_tus.cpp
  SYCL :: Assert/assert_in_multiple_tus_one_ndebug.cpp
  SYCL :: Assert/assert_in_one_kernel.cpp
  SYCL :: Assert/assert_in_simultaneous_kernels.cpp
  SYCL :: Assert/assert_in_simultaneously_multiple_tus.cpp
  SYCL :: Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp
  SYCL :: BFloat16/bfloat16_builtins.cpp
  SYCL :: Matrix/element_wise_all_ops_cuda.cpp
  SYCL :: Matrix/element_wise_wi_marray.cpp
  SYCL :: Matrix/joint_matrix_tensorcore.cpp
``` 

It somewhat overlaps. Tried #1072 but it doesn't help.

**Comment by JackAKirk** [comment]
> @JackAKirk, @steffenlarsen, I couldn't reproduce on other cuda machine, but when ran on the runner got a bunch of issues:
> 
> ```
> Failed Tests (11):
>   SYCL :: Assert/assert_in_kernels.cpp
>   SYCL :: Assert/assert_in_multiple_tus.cpp
>   SYCL :: Assert/assert_in_multiple_tus_one_ndebug.cpp
>   SYCL :: Assert/assert_in_one_kernel.cpp
>   SYCL :: Assert/assert_in_simultaneous_kernels.cpp
>   SYCL :: Assert/assert_in_simultaneously_multiple_tus.cpp
>   SYCL :: Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp
>   SYCL :: BFloat16/bfloat16_builtins.cpp
>   SYCL :: Matrix/element_wise_all_ops_cuda.cpp
>   SYCL :: Matrix/element_wise_wi_marray.cpp
>   SYCL :: Matrix/joint_matrix_tensorcore.cpp
> ```
> 
> It somewhat overlaps. Tried #1072 but it doesn't help.



> @JackAKirk, @steffenlarsen, I couldn't reproduce on other cuda machine, but when ran on the runner got a bunch of issues:
> 
> ```
> Failed Tests (11):
>   SYCL :: Assert/assert_in_kernels.cpp
>   SYCL :: Assert/assert_in_multiple_tus.cpp
>   SYCL :: Assert/assert_in_multiple_tus_one_ndebug.cpp
>   SYCL :: Assert/assert_in_one_kernel.cpp
>   SYCL :: Assert/assert_in_simultaneous_kernels.cpp
>   SYCL :: Assert/assert_in_simultaneously_multiple_tus.cpp
>   SYCL :: Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp
>   SYCL :: BFloat16/bfloat16_builtins.cpp
>   SYCL :: Matrix/element_wise_all_ops_cuda.cpp
>   SYCL :: Matrix/element_wise_wi_marray.cpp
>   SYCL :: Matrix/joint_matrix_tensorcore.cpp
> ```
> 
> It somewhat overlaps. Tried #1072 but it doesn't help.

@pvchupin The errors reported by the CI in the bfloat16 tests are all due to experimental headers not being included correctly: https://github.com/intel/llvm/pull/6386 might lead to different behaviour since the tests will have to include the experimental/builtins.hpp header explicitly instead of relying on sycl.hpp: I think I should have made this change already since apparently no C++17 usage should be included in sycl.hpp: although this did not lead to failing tests previously.

**Comment by pvchupin** [comment]
@JackAKirk, I think it is OK to have C++17 usage in the sycl.hpp **if** it's guarded and excluded properly, so that C++14 host compilers don't complain. It's ok to have new/experimental features missing in C++14 mode.

To summarize:
* C++17 is SYCL2020 default and we'd like to be free in writing C++17 code
* Some customers lagging with move from C++14 forward, so for them we guard C++17 features so that nothing existing breaks on their side.

Not sure if that changes any of the patches. I'm testing these now and will merge if everything works (just to stop confusion in other PRs).

**Comment by JackAKirk** [comment]
> @JackAKirk, I think it is OK to have C++17 usage in the sycl.hpp **if** it's guarded and excluded properly, so that C++14 host compilers don't complain. It's ok to have new/experimental features missing in C++14 mode.
> 
> To summarize:
> 
>     * C++17 is SYCL2020 default and we'd like to be free in writing C++17 code
> 
>     * Some customers lagging with move from C++14 forward, so for them we guard C++17 features so that nothing existing breaks on their side.
> 
> 
> Not sure if that changes any of the patches. I'm testing these now and will merge if everything works (just to stop confusion in other PRs).

I see, thanks for clarifying that.

**Pull request merged**
